File: priv/zig/robot_race_domain.zig
const std = @import("std");
const beam = @import("beam");

const ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const ID_SIZE: usize = 5;

const Role = enum { guest, admin };
const State = enum { setup, counting_down, playing, finished };

const Range = struct {
    first: u64,
    last: u64,
    step: i64,
};

const GameConfig = struct {
    winning_score: u64,
    num_robots: Range,
    countdown: u64,
};

const Robot = struct {
    id: []const u8,
    name: []const u8,
    role: Role,
    score: u64,
};

const PreviousWin = struct {
    id: []const u8,
    wins: u64,
};

const Game = struct {
    id: []const u8,
    winning_score: u64,
    num_robots: Range,
    countdown: u64,
    config: GameConfig,
    robots: []Robot,
    state: State,
    previous_wins: []PreviousWin,
};

const Stats = struct {
    num_games: u64,
};

const JoinResult = union(enum) {
    ok: Game,
    game_in_progress,
    game_full,
};

const LeaderboardRow = struct {
    robot: Robot,
    win_count: u64,
};

const NewRobotArgs = struct {
    name: []const u8,
    role: Role,
};

const NewGameArgs = struct {
    config: GameConfig,
};

const JoinArgs = struct {
    game: Game,
    robot: Robot,
};

const ScorePointArgs = struct {
    game: Game,
    robot_id: []const u8,
};

const AdminArgs = struct {
    game: Game,
    robot_id: []const u8,
};

const GameArgs = struct {
    game: Game,
};

const IncrementNumGamesArgs = struct {
    stats: Stats,
};

pub fn call_nif(function_name: []const u8, args_term: beam.term) !beam.term {
    var arena_state = std.heap.ArenaAllocator.init(beam.allocator);
    defer arena_state.deinit();
    const allocator = arena_state.allocator();

    if (std.mem.eql(u8, function_name, "new_id")) {
        return beam.make(try newId(allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "new_game_id")) {
        return beam.make(try newPrefixedId("g_", allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "new_robot_id")) {
        return beam.make(try newPrefixedId("r_", allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "new_robot")) {
        const args = try beam.get(NewRobotArgs, args_term, .{ .allocator = allocator });
        const robot = try newRobot(args.name, args.role, allocator);
        return beam.make(robot, .{});
    }

    if (std.mem.eql(u8, function_name, "new_stats")) {
        return beam.make(Stats{ .num_games = 0 }, .{});
    }

    if (std.mem.eql(u8, function_name, "increment_num_games")) {
        const args = try beam.get(IncrementNumGamesArgs, args_term, .{ .allocator = allocator });

        var stats = args.stats;
        stats.num_games += 1;

        return beam.make(stats, .{});
    }

    if (std.mem.eql(u8, function_name, "new_game")) {
        const args = try beam.get(NewGameArgs, args_term, .{ .allocator = allocator });
        return beam.make(try newGame(args.config, allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "join")) {
        const args = try beam.get(JoinArgs, args_term, .{ .allocator = allocator });

        switch (try join(args.game, args.robot, allocator)) {
            .ok => |game| return beam.make(.{ .status = .ok, .game = game }, .{}),
            .game_in_progress => return beam.make(.{ .status = .failed, .reason = .game_in_progress }, .{}),
            .game_full => return beam.make(.{ .status = .failed, .reason = .game_full }, .{}),
        }
    }

    if (std.mem.eql(u8, function_name, "score_point")) {
        const args = try beam.get(ScorePointArgs, args_term, .{ .allocator = allocator });
        return beam.make(try scorePoint(args.game, args.robot_id, allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "robots")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });
        return beam.make(args.game.robots, .{});
    }

    if (std.mem.eql(u8, function_name, "admin")) {
        const args = try beam.get(AdminArgs, args_term, .{ .allocator = allocator });
        return beam.make(admin(args.game, args.robot_id), .{});
    }

    if (std.mem.eql(u8, function_name, "play")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });

        var game = args.game;
        game.state = .playing;

        return beam.make(game, .{});
    }

    if (std.mem.eql(u8, function_name, "countdown")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });
        return beam.make(countdown(args.game), .{});
    }

    if (std.mem.eql(u8, function_name, "score_board")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });
        return beam.make(try scoreBoard(args.game, allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "play_again")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });
        return beam.make(try playAgain(args.game, allocator), .{});
    }

    if (std.mem.eql(u8, function_name, "leaderboard")) {
        const args = try beam.get(GameArgs, args_term, .{ .allocator = allocator });
        return beam.make(try leaderboard(args.game, allocator), .{});
    }

    return error.badarg;
}

fn newId(allocator: std.mem.Allocator) ![]u8 {
    var id = try allocator.alloc(u8, ID_SIZE);

    var random_bytes: [ID_SIZE]u8 = undefined;
    std.crypto.random.bytes(&random_bytes);

    for (random_bytes, 0..) |byte, idx| {
        id[idx] = ALPHABET[@as(usize, byte) % ALPHABET.len];
    }

    return id;
}

fn newPrefixedId(prefix: []const u8, allocator: std.mem.Allocator) ![]u8 {
    const suffix = try newId(allocator);
    var id = try allocator.alloc(u8, prefix.len + suffix.len);

    for (prefix, 0..) |char, idx| {
        id[idx] = char;
    }

    for (suffix, 0..) |char, idx| {
        id[prefix.len + idx] = char;
    }

    return id;
}

fn newRobot(name: []const u8, role: Role, allocator: std.mem.Allocator) !Robot {
    return Robot{
        .id = try newPrefixedId("r_", allocator),
        .name = name,
        .role = role,
        .score = 0,
    };
}

fn newGame(config: GameConfig, allocator: std.mem.Allocator) !Game {
    return Game{
        .id = try newPrefixedId("g_", allocator),
        .winning_score = config.winning_score,
        .num_robots = config.num_robots,
        .countdown = config.countdown,
        .config = config,
        .robots = try allocator.alloc(Robot, 0),
        .state = .setup,
        .previous_wins = try allocator.alloc(PreviousWin, 0),
    };
}

fn join(game: Game, robot: Robot, allocator: std.mem.Allocator) !JoinResult {
    if (game.state == .counting_down or game.state == .playing or game.state == .finished) {
        return .game_in_progress;
    }

    const max_robots: usize = @intCast(game.num_robots.last);

    if (game.robots.len >= max_robots) {
        return .game_full;
    }

    var updated_game = game;
    var robots = try allocator.alloc(Robot, game.robots.len + 1);

    for (game.robots, 0..) |existing_robot, idx| {
        robots[idx] = existing_robot;
    }

    robots[game.robots.len] = robot;
    updated_game.robots = robots;

    return .{ .ok = updated_game };
}

fn scorePoint(game: Game, robot_id: []const u8, allocator: std.mem.Allocator) !Game {
    if (game.state != .playing) {
        return game;
    }

    var updated_game = game;
    updated_game.robots = try cloneRobots(game.robots, allocator);

    var matched: ?usize = null;

    for (updated_game.robots, 0..) |*robot, idx| {
        if (std.mem.eql(u8, robot.id, robot_id)) {
            robot.score += 1;
            matched = idx;
            break;
        }
    }

    if (matched) |idx| {
        if (updated_game.robots[idx].score >= updated_game.winning_score) {
            updated_game.state = .finished;
        }
    }

    return updated_game;
}

fn admin(game: Game, robot_id: []const u8) ?bool {
    for (game.robots) |robot| {
        if (std.mem.eql(u8, robot.id, robot_id)) {
            return robot.role == .admin;
        }
    }

    return null;
}

fn countdown(game: Game) Game {
    var updated_game = game;

    if (game.state == .setup) {
        updated_game.state = .counting_down;
        return updated_game;
    }

    if (game.countdown > 0) {
        updated_game.state = .counting_down;
        updated_game.countdown = game.countdown - 1;
        return updated_game;
    }

    if (game.countdown == 0) {
        updated_game.state = .playing;
    }

    return updated_game;
}

fn scoreBoard(game: Game, allocator: std.mem.Allocator) ![]Robot {
    const board = try cloneRobots(game.robots, allocator);
    sortRobotsByScoreDesc(board);
    return board;
}

fn playAgain(game: Game, allocator: std.mem.Allocator) !Game {
    const winner_robot = try winner(game, allocator);

    var updated_game = game;
    updated_game.winning_score = game.config.winning_score;
    updated_game.num_robots = game.config.num_robots;
    updated_game.countdown = game.config.countdown;
    updated_game.robots = try cloneRobots(game.robots, allocator);

    for (updated_game.robots) |*robot| {
        robot.score = 0;
    }

    updated_game.state = .setup;
    updated_game.previous_wins = try incrementPreviousWins(game.previous_wins, winner_robot.id, allocator);

    return updated_game;
}

fn leaderboard(game: Game, allocator: std.mem.Allocator) ![]LeaderboardRow {
    const winner_robot = try winner(game, allocator);

    var rows = try allocator.alloc(LeaderboardRow, game.robots.len);

    for (game.robots, 0..) |robot, idx| {
        const previous_win_count = previousWinsCount(game.previous_wins, robot.id);
        const win_count = if (std.mem.eql(u8, robot.id, winner_robot.id)) previous_win_count + 1 else previous_win_count;

        rows[idx] = .{ .robot = robot, .win_count = win_count };
    }

    sortLeaderboardRowsByWinCountDesc(rows);

    return rows;
}

fn winner(game: Game, allocator: std.mem.Allocator) !Robot {
    const board = try scoreBoard(game, allocator);

    if (board.len == 0) {
        return error.empty_game;
    }

    return board[0];
}

fn cloneRobots(robots: []const Robot, allocator: std.mem.Allocator) ![]Robot {
    var cloned = try allocator.alloc(Robot, robots.len);

    for (robots, 0..) |robot, idx| {
        cloned[idx] = robot;
    }

    return cloned;
}

fn incrementPreviousWins(previous_wins: []const PreviousWin, winner_id: []const u8, allocator: std.mem.Allocator) ![]PreviousWin {
    var updated = try allocator.alloc(PreviousWin, previous_wins.len);

    for (previous_wins, 0..) |entry, idx| {
        updated[idx] = entry;
    }

    for (updated) |*entry| {
        if (std.mem.eql(u8, entry.id, winner_id)) {
            entry.wins += 1;
            return updated;
        }
    }

    var expanded = try allocator.alloc(PreviousWin, previous_wins.len + 1);

    for (updated, 0..) |entry, idx| {
        expanded[idx] = entry;
    }

    expanded[previous_wins.len] = .{ .id = winner_id, .wins = 1 };

    return expanded;
}

fn previousWinsCount(previous_wins: []const PreviousWin, robot_id: []const u8) u64 {
    for (previous_wins) |entry| {
        if (std.mem.eql(u8, entry.id, robot_id)) {
            return entry.wins;
        }
    }

    return 0;
}

fn sortRobotsByScoreDesc(robots: []Robot) void {
    var i: usize = 0;

    while (i < robots.len) : (i += 1) {
        var j: usize = i + 1;

        while (j < robots.len) : (j += 1) {
            if (robots[j].score > robots[i].score) {
                std.mem.swap(Robot, &robots[i], &robots[j]);
            }
        }
    }
}

fn sortLeaderboardRowsByWinCountDesc(rows: []LeaderboardRow) void {
    var i: usize = 0;

    while (i < rows.len) : (i += 1) {
        var j: usize = i + 1;

        while (j < rows.len) : (j += 1) {
            if (rows[j].win_count > rows[i].win_count) {
                std.mem.swap(LeaderboardRow, &rows[i], &rows[j]);
            }
        }
    }
}

File: priv/python/robot_race_domain.py
import copy
import json
import os
import secrets
import threading
from datetime import datetime, timezone
from pathlib import Path

ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
ID_SIZE = 5
IN_PROGRESS_STATES = {"counting_down", "playing", "finished"}
DEFAULT_LOG_PATH = Path(__file__).with_name("robot_race_domain.log")
LOG_PATH = Path(os.environ.get("ROBOT_RACE_DOMAIN_LOG_PATH", str(DEFAULT_LOG_PATH)))
LOG_LOCK = threading.Lock()


def _log(event, **fields):
    entry = {
        "timestamp": datetime.now(timezone.utc).isoformat(timespec="milliseconds"),
        "event": event,
        **fields,
    }

    try:
        with LOG_LOCK:
            with LOG_PATH.open("a", encoding="utf-8") as log_file:
                log_file.write(json.dumps(entry, sort_keys=True))
                log_file.write("\n")
    except Exception:
        # Never fail game operations because of logging.
        pass


def new_id(_args=None):
    return "".join(secrets.choice(ALPHABET) for _ in range(ID_SIZE))


def new_game_id(_args=None):
    return "g_" + new_id()


def new_robot_id(_args=None):
    return "r_" + new_id()


def new_robot(args):
    robot = {
        "id": new_robot_id(),
        "name": args["name"],
        "role": args["role"],
        "score": 0,
    }

    _log("new_robot", robot_id=robot["id"], name=robot["name"], role=robot["role"])
    return robot


def new_stats(_args=None):
    return {"num_games": 0}


def increment_num_games(args):
    stats = copy.deepcopy(args["stats"])
    before = stats["num_games"]
    stats["num_games"] = stats["num_games"] + 1
    _log("increment_num_games", num_games_before=before, num_games_after=stats["num_games"])
    return stats


def new_game(args):
    config = copy.deepcopy(args["config"])

    game = {
        "id": new_game_id(),
        "winning_score": config["winning_score"],
        "num_robots": config["num_robots"],
        "countdown": config["countdown"],
        "config": config,
        "robots": [],
        "state": "setup",
        "previous_wins": {},
    }

    _log(
        "new_game",
        game_id=game["id"],
        winning_score=game["winning_score"],
        countdown=game["countdown"],
        min_robots=game["num_robots"]["first"],
        max_robots=game["num_robots"]["last"],
    )

    return game


def join(args):
    game = copy.deepcopy(args["game"])
    robot = copy.deepcopy(args["robot"])
    before_count = len(game["robots"])

    if game["state"] in IN_PROGRESS_STATES:
        _log(
            "join_rejected",
            game_id=game["id"],
            robot_id=robot["id"],
            robot_name=robot["name"],
            reason="game_in_progress",
            state=game["state"],
            robot_count=before_count,
        )
        return {"status": "error", "error": "game_in_progress"}

    max_robots = game["num_robots"]["last"]
    if len(game["robots"]) >= max_robots:
        _log(
            "join_rejected",
            game_id=game["id"],
            robot_id=robot["id"],
            robot_name=robot["name"],
            reason="game_full",
            state=game["state"],
            robot_count=before_count,
            max_robots=max_robots,
        )
        return {"status": "error", "error": "game_full"}

    game["robots"].append(robot)
    _log(
        "join_ok",
        game_id=game["id"],
        robot_id=robot["id"],
        robot_name=robot["name"],
        robot_count=len(game["robots"]),
    )
    return {"status": "ok", "game": game}


def score_point(args):
    game = copy.deepcopy(args["game"])
    robot_id = args["robot_id"]

    if game["state"] != "playing":
        _log(
            "score_point_ignored",
            game_id=game["id"],
            robot_id=robot_id,
            state=game["state"],
        )
        return game

    updated_robot = None
    before_score = None

    for robot in game["robots"]:
        if robot["id"] == robot_id:
            before_score = robot["score"]
            robot["score"] = robot["score"] + 1
            updated_robot = robot
            break

    if updated_robot is not None and updated_robot["score"] >= game["winning_score"]:
        game["state"] = "finished"

    if updated_robot is None:
        _log(
            "score_point_unknown_robot",
            game_id=game["id"],
            robot_id=robot_id,
            state=game["state"],
        )
    else:
        _log(
            "score_point",
            game_id=game["id"],
            robot_id=robot_id,
            before_score=before_score,
            after_score=updated_robot["score"],
            state=game["state"],
        )

    return game


def robots(args):
    game = args["game"]
    return copy.deepcopy(game["robots"])


def admin(args):
    game = args["game"]
    robot_id = args["robot_id"]

    for robot in game["robots"]:
        if robot["id"] == robot_id:
            return robot["role"] == "admin"

    return None


def play(args):
    game = copy.deepcopy(args["game"])
    before_state = game["state"]
    game["state"] = "playing"
    _log("play", game_id=game["id"], state_before=before_state, state_after=game["state"])
    return game


def countdown(args):
    game = copy.deepcopy(args["game"])
    before_state = game["state"]
    before_countdown = game["countdown"]

    if game["state"] == "setup":
        game["state"] = "counting_down"
        _log(
            "countdown",
            game_id=game["id"],
            state_before=before_state,
            state_after=game["state"],
            countdown_before=before_countdown,
            countdown_after=game["countdown"],
        )
        return game

    if game["countdown"] > 0:
        game["state"] = "counting_down"
        game["countdown"] = game["countdown"] - 1
        _log(
            "countdown",
            game_id=game["id"],
            state_before=before_state,
            state_after=game["state"],
            countdown_before=before_countdown,
            countdown_after=game["countdown"],
        )
        return game

    if game["countdown"] == 0:
        game["state"] = "playing"

    _log(
        "countdown",
        game_id=game["id"],
        state_before=before_state,
        state_after=game["state"],
        countdown_before=before_countdown,
        countdown_after=game["countdown"],
    )

    return game


def score_board(args):
    game = args["game"]
    return sorted(copy.deepcopy(game["robots"]), key=lambda robot: robot["score"], reverse=True)


def winner(args):
    board = score_board({"game": args["game"]})
    if not board:
        raise IndexError("winner called for an empty scoreboard")
    return board[0]


def play_again(args):
    game = copy.deepcopy(args["game"])
    before_state = game["state"]
    winner_robot = winner({"game": game})

    game["winning_score"] = game["config"]["winning_score"]
    game["num_robots"] = copy.deepcopy(game["config"]["num_robots"])
    game["countdown"] = game["config"]["countdown"]
    game["robots"] = [reset_robot_score(robot) for robot in game["robots"]]
    game["state"] = "setup"
    game["previous_wins"] = save_winner(game["previous_wins"], winner_robot["id"])

    _log(
        "play_again",
        game_id=game["id"],
        winner_robot_id=winner_robot["id"],
        state_before=before_state,
        state_after=game["state"],
        countdown_reset=game["countdown"],
        winner_total_wins=game["previous_wins"].get(winner_robot["id"], 0),
    )

    return game


def leaderboard(args):
    game = copy.deepcopy(args["game"])
    current_winner_id = winner({"game": game})["id"]

    leaderboard_rows = []
    for robot in game["robots"]:
        previous_win_count = game["previous_wins"].get(robot["id"], 0)

        if current_winner_id == robot["id"]:
            win_count = previous_win_count + 1
        else:
            win_count = previous_win_count

        leaderboard_rows.append((robot, win_count))

    leaderboard_rows.sort(key=lambda row: row[1], reverse=True)

    return leaderboard_rows


def reset_robot_score(robot):
    updated_robot = copy.deepcopy(robot)
    updated_robot["score"] = 0
    return updated_robot


def save_winner(previous_wins, winner_id):
    updated_previous_wins = copy.deepcopy(previous_wins)
    updated_previous_wins[winner_id] = updated_previous_wins.get(winner_id, 0) + 1
    return updated_previous_wins
